### PR TITLE
Add a pass to calculate machine function's cfg hash to detect whether…

### PR DESCRIPTION
**Pull request by lifengxiang1025**
… Propeller's profile is out-dated [1/2]

### CODE REVIEW

**Comment by lifengxiang1025** [comment]
Hi @rlavaee 
Please review this patch. I plan to add changes to AsmPrinter/objdump/readobj in part2 when this patch is adopted. Thanks.

**Comment by lifengxiang1025** [comment]
> C/C++ code formatter, clang-format found issues in your code.
> -void initializeXRayInstrumentationPass(PassRegistry&);
> +void initializeXRayInstrumentationPass(PassRegistry &);
It confuses me. I don't change this code.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+using namespace llvm;
+static cl::opt<bool>
+    MFCFGHashDump("mf-cfg-hash-dump",
+                  cl::desc("Dump machine function's control flow grpah hash"),
+                  cl::init(false), cl::Hidden);
+
+// Calculate machine function hash in level order traversal.
+// For each machine basic block, using its mbb's BaseID,
+// size of successors and  successors' mbb's BaseID to update hash.
+// These informations can make graph unique.
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:33]: This information

Comment (rlavaee) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:33]: Although I think this sentence is a bit ambiguous. Maybe rephrase to say "The control flow graph is uniquely represented by its level-order traversal". BTW, do you enforce any order on the traversal of the successors? If you don't, it won't be unique.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        Now->getBBID()->BaseID);
+    uint32_t Size =
+        endian::byte_swap<uint32_t, llvm::endianness::little>(Now->succ_size());
+    Hash.update(llvm::ArrayRef((uint8_t *)&Value, sizeof(Value)));
+    Hash.update(llvm::ArrayRef((uint8_t *)&Size, sizeof(Size)));
+    if (Visited.count(Now)) {
+      continue;
+    }
+    Visited.insert(Now);
+    for (MachineBasicBlock *Succ : Now->successors()) {
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:55]: You may want to enforce an ordering here. Maybe based on `BaseID`.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+static cl::opt<bool>
+    MFCFGHashDump("mf-cfg-hash-dump",
+                  cl::desc("Dump machine function's control flow grpah hash"),
+                  cl::init(false), cl::Hidden);
+
+// Calculate machine function hash in level order traversal.
+// For each machine basic block, using its mbb's BaseID,
+// size of successors and  successors' mbb's BaseID to update hash.
+// These informations can make graph unique.
+static uint64_t calculateMBBCFGHash(MachineFunction &MF) {
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:34]: This can be `const MachineFunction &`. Right?

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  std::unordered_set<MachineBasicBlock *> Visited;
+  MD5 Hash;
+  std::queue<MachineBasicBlock *> Q;
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:37]: Maybe change to:
if (MF.empty()) return 0;
std::queue<const MachineBasicBlock *> WorkList(&MF.front());
...

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  if (!MF.empty()) {
+    Q.push(&*MF.begin());
+  }
+  while (!Q.empty()) {
+    MachineBasicBlock *Now = Q.front();
+    Q.pop();
+    using namespace llvm::support;
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:44]: Just use `support::endian::byte_swap`.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    uint32_t Value = endian::byte_swap<uint32_t, llvm::endianness::little>(
+        Now->getBBID()->BaseID);
+    uint32_t Size =
+        endian::byte_swap<uint32_t, llvm::endianness::little>(Now->succ_size());
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:48]: No need to use `llvm::` here. You already have `using namespace llvm` in this file.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    Hash.update(llvm::ArrayRef((uint8_t *)&Value, sizeof(Value)));
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:49]: same here.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+// These informations can make graph unique.
+static uint64_t calculateMBBCFGHash(MachineFunction &MF) {
+  std::unordered_set<MachineBasicBlock *> Visited;
+  MD5 Hash;
+  std::queue<MachineBasicBlock *> Q;
+  if (!MF.empty()) {
+    Q.push(&*MF.begin());
+  }
+  while (!Q.empty()) {
+    MachineBasicBlock *Now = Q.front();
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:42]: `CurrentBB` or `Current` is a better name here.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    Q.pop();
+    using namespace llvm::support;
+    uint32_t Value = endian::byte_swap<uint32_t, llvm::endianness::little>(
+        Now->getBBID()->BaseID);
+    uint32_t Size =
+        endian::byte_swap<uint32_t, llvm::endianness::little>(Now->succ_size());
+    Hash.update(llvm::ArrayRef((uint8_t *)&Value, sizeof(Value)));
+    Hash.update(llvm::ArrayRef((uint8_t *)&Size, sizeof(Size)));
+    if (Visited.count(Now)) {
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:51]: Remove braces for single-line conditionals.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  }
+  llvm::MD5::MD5Result Result;
+  Hash.final(Result);
+  return Result.low();
+}
+
+bool MachineFunctionHashBuilder::runOnMachineFunction(MachineFunction &MF) {
+  setCFGHash(MF.getName(), calculateMBBCFGHash(MF));
+  if (MFCFGHashDump) {
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:66]: How important is this option? Maybe we can remove it.

File: llvm/lib/CodeGen/TargetPassConfig.cpp [hunk 11]
```diff
@@ -254,6 +254,11 @@ static cl::opt<bool>
     GCEmptyBlocks("gc-empty-basic-blocks", cl::init(false), cl::Hidden,
                   cl::desc("Enable garbage-collecting empty basic blocks"));
 
+/// Enable calculating machine function's cfg hash.
+static cl::opt<bool> UseMachineFunctionCFGHash(
+    "mf-cfg-hash", cl::init(false), cl::Hidden,
+    cl::desc("Enable calculating machine function's cfg hash."));
```
Comment (rlavaee) [inline llvm/lib/CodeGen/TargetPassConfig.cpp:260]: Could you please rename and rephrase the description of this option? It doesn't mention what it will be used for.

File: llvm/test/CodeGen/X86/machine-function-cfg-hash.ll [hunk 12]
```diff
@@ -0,0 +1,81 @@
+; RUN: llc < %s -O0 -mtriple=x86_64-pc-linux -function-sections -basic-block-sections=labels \
```
Comment (rlavaee) [inline llvm/test/CodeGen/X86/machine-function-cfg-hash.ll:1]: Maybe we don't need to add and test the dump option.

File: llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesAll();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  uint64_t getCFGHash(StringRef Name) {
+    if (!CFGHash.count(Name)) {
```
Comment (rlavaee) [inline llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h:45]: `return CFGHash.lookup(Name);`

File: llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesAll();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  uint64_t getCFGHash(StringRef Name) {
```
Comment (rlavaee) [inline llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h:44]: Add a function comment.

**Review by rlavaee — COMMENTED** [review]
Sorry for the delay. A first round of review.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 2 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Thanks. Done.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 3 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 4 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 5 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 6 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 7 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 8 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 9 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp (hunk 10 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp]: Done. Remove it.

File: llvm/test/CodeGen/X86/machine-function-cfg-hash.ll (hunk 12 above)
Comment (lifengxiang1025) [reply llvm/test/CodeGen/X86/machine-function-cfg-hash.ll]: Done

File: llvm/lib/CodeGen/TargetPassConfig.cpp (hunk 11 above)
Comment (lifengxiang1025) [reply llvm/lib/CodeGen/TargetPassConfig.cpp]: Done.

File: llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h (hunk 13 above)
Comment (lifengxiang1025) [reply llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h]: Done

File: llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h (hunk 14 above)
Comment (lifengxiang1025) [reply llvm/include/llvm/CodeGen/MachineFunctionHashBuilder.h]: Done

**Comment by lifengxiang1025** [comment]
> Sorry for the delay. A first round of review.

Thanks! I have updated code based on your code review.

File: llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+  std::queue<const MachineBasicBlock *> WorkList;
+  WorkList.push(&*MF.begin());
+  while (!WorkList.empty()) {
+    const MachineBasicBlock *CurrentBB = WorkList.front();
+    WorkList.pop();
+    uint32_t Value = support::endian::byte_swap<uint32_t, endianness::little>(
+        CurrentBB->getBBID()->BaseID);
+    uint32_t Size = support::endian::byte_swap<uint32_t, endianness::little>(
+        CurrentBB->succ_size());
+    Hash.update(ArrayRef((uint8_t *)&Value, sizeof(Value)));
```
Comment (rlavaee) [inline llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:42]: You may update the hash several times with the same element. A single item may be inserted and visited several times since `Visited.count(CurrentBB)` is checked later. I think this is unnecessary. You can do this check before pushing items in the worklist. Also, please initialize `Visited` with the entry block.

Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:42]: For example, the cfg is:
`0->2->3`
`  ->1->3 `
I want to restore the cfg by BB's id and BB's successor's size. So the sequence will be 
`0(id) 2(successor size) 1(id) 1(successor size) 2(id) 1(successor size) 3(id) 0(successor size) 3(id) 0(successor size)`.

Comment (tmsri) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:42]: I don't fully understand the various cases you are trying to tolerate with this hashing.  You are only hashing the size and the id of the block.   What are the cases you are trying to capture here?  Why not hash the contents of the block?

In this mechanism, the cfg hashing will not tolerate a change to the compiler where the blocks had the same CFG but somehow got re-numbered, right?

Comment (lifengxiang1025) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:42]: > In this mechanism, the cfg hashing will not tolerate a change to the compiler where the blocks had the same CFG but somehow got re-numbered, right?

Thanks for reply. Yes, you are right.
I think we can divide cfg hashing to three version: loose hash, strict hash, full hash which Bolt do. For example, how about the strict hash is calculated by the size and the id of the block and the full hash is calculated by the contents of the function? More strict hash will cause more compile time.

Comment (tmsri) [reply llvm/lib/CodeGen/MachineFunctionHashBuilder.cpp:42]: Can you measure the additional overhead of doing a full/strict hash over a loose hash?
